### PR TITLE
Firestore/favourites

### PR DESCRIPTION
**Pull request by quiro91**
## Problem

We were not able to add, remove or read any user favorite event

## Solution

Implement the persistance of the favourites in Firestore

### Test(s) added 

@rock3r added a couple regarding the schedule

### Paired with 

Some tests and merging from @rock3r, but mostly alone

### CODE REVIEW

File: app/src/main/java/net/squanchy/favorites/view/FavoritesAdapter.kt [hunk 1]
```diff
@@ -84,7 +84,7 @@ internal class FavoritesAdapter(context: Context?) : RecyclerView.Adapter<Recycl
                 0,
                 position,
                 schedule.pages,
-                { schedule.pages[position].date },
+                { it.date },
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/favorites/view/FavoritesAdapter.kt]: maybe rename it to `page` for readability

File: app/src/main/java/net/squanchy/schedule/ScheduleService.kt [hunk 2]
```diff
@@ -54,10 +55,17 @@ class FirestoreScheduleService(
...
+        return Observable.combineLatest(
+            this,
+            authService.ifUserSignedInThenObservableFrom(dbService::favorites),
+            BiFunction { schedule: List<FirestoreSchedulePage>, favorites: List<FirestoreFavorite> ->
+                schedule.filterPagesEvents { favorites.includes(it) }
+            })
+    }
 
-    private fun FirestoreEvent.isFavorite() = false // TODO add actual favourites filtering
+    private fun List<FirestoreFavorite>.includes(event: FirestoreEvent) =
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/ScheduleService.kt]: Can you only pass in the `eventId` if that's all you need?

File: app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt [hunk 3]
```diff
@@ -11,7 +11,7 @@ private const val CURRENT_SLOT_THRESHOLD = .6f
 data class Schedule(val pages: List<SchedulePage>, val timeZone: DateTimeZone) {
 
     val isEmpty: Boolean
-        get() = pages.isEmpty()
+        get() = pages.all { it.events.isEmpty() }
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/domain/view/Schedule.kt:14]: Best fix ever 💃

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt [hunk 4]
```diff
@@ -141,9 +145,9 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
...
         return Observable.create { subscriber ->
-            val registration = db.collection("views")
-                .document("event_details")
-                .collection("events")
+            val registration = db.collection(VIEWS)
+                .document(EVENT_DETAILS)
+                .collection(EVENTS)
                 .document(eventId)
                 .addSnapshotListener { snapshot, exception ->
                     if (exception != null && subscriber.isDisposed.not()) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: No love for Tracks at line 166? 😢 

... I'll take care of it with the PR switching to using the tracks view, no need to do it now.

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt [hunk 5]
```diff
@@ -17,10 +21,10 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
 
     fun scheduleView(): Observable<List<FirestoreSchedulePage>> {
         return Observable.create { subscriber ->
-            val registration = db.collection("views")
-                .document("schedule")
-                .collection("schedule_pages")
-                .orderBy("day.date")
+            val registration = db.collection(VIEWS)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: maybe we could extract `db.collection(VIEWS).document(SCHEDULE)` as a method for brevity:

```kotlin
fun FirestoreDbService.view(viewName: String) = db.collection(VIEWS).document(viewName)
```

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt [hunk 6]
```diff
@@ -173,4 +177,73 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
...
+                    subscriber.onNext(snapshot.documents.map { snapshot ->
+                        snapshot.toObject(FirestoreFavorite::class.java)
+                    })
+                }
+
+            subscriber.setCancellable(registration::remove)
+        }
+    }
+
+    private val addToFirestoreAction: (DocumentReference.(String) -> Task<*>) = { eventId: String ->
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: This should go _after_ `addFavorite()`

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt [hunk 7]
```diff
@@ -173,4 +177,73 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
...
+        val firestoreFavorite = FirestoreFavorite().apply { id = eventId }
+        collection(FAVORITES).document(eventId).set(firestoreFavorite)
+    }
+
+    private val deleteFromFirestoreAction: (DocumentReference.(String) -> Task<*>) = { eventId ->
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: This should go _after_ `removeFavorite()`

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt [hunk 8]
```diff
@@ -173,4 +177,73 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
...
+        return Observable.create { subscriber ->
+            val registration = db.collection(USER_DATA)
+                .document(userId)
+                .collection(FAVORITES)
+                .addSnapshotListener { snapshot, exception ->
+                    if (exception != null && subscriber.isDisposed.not()) {
+                        subscriber.onError(exception)
+                        return@addSnapshotListener
+                    }
+                    subscriber.onNext(snapshot.documents.map { snapshot ->
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: ⚠️ `snapshot` shadows the one at line 186, rename to `favoriteSnapshot`

File: app/src/main/java/net/squanchy/service/repository/firebase/FirestoreEventRepository.kt [hunk 9]
```diff
@@ -2,59 +2,61 @@ package net.squanchy.service.repository.firebase
...
-    private fun FirestoreVenue.extractTimeZone(): DateTimeZone {
-        return DateTimeZone.forID(timezone)
-    }
+    private fun FirestoreVenue.extractTimeZone() = DateTimeZone.forID(timezone)
+
+    private fun combineIntoEvent(event: FirestoreEvent, timeZone: DateTimeZone, favorites: List<FirestoreFavorite>): Event =
+        event.toEvent(checksum, timeZone, favorites.includes(event))
 
-    private fun combineIntoEvent(event: FirestoreEvent, timeZone: DateTimeZone): Event = event.toEvent(checksum, timeZone)
+    private fun combineIntoEvents(events: List<FirestoreEvent>, timeZone: DateTimeZone, favorites: List<FirestoreFavorite>): List<Event> =
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirestoreEventRepository.kt]: Probably my bad, but this should go after `override fun events()`, and `combineIntoEvent()` should go after this (top-down order)

File: app/src/main/java/net/squanchy/service/repository/firebase/FirestoreEventRepository.kt [hunk 10]
```diff
@@ -2,59 +2,61 @@ package net.squanchy.service.repository.firebase
...
+        events.map { combineIntoEvent(it, timeZone, favorites) }
 
-    private fun combineIntoEvents(events: List<FirestoreEvent>, timeZone: DateTimeZone): List<Event> = events.map { combineIntoEvent(it, timeZone) }
+    private fun List<FirestoreFavorite>.includes(event: FirestoreEvent) =
+        mapNotNull { it.id }.contains(event.id)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirestoreEventRepository.kt]: Can we do the mapping only once and pass in here a list of IDs? This is an extra step we do for each event in the schedule and seems wasteful.

File: app/src/test/java/net/squanchy/schedule/FirestoreScheduleServiceTest.kt [hunk 11]
```diff
@@ -183,7 +185,33 @@ class FirestoreScheduleServiceTest {
             )
     }
 
-    // TODO test events are filtered out when onyFavorites and they're not favorited
+    @Test
+    fun `should exclude events that are not favorites when favorite is true`() {
```
Comment (rock3r) [inline app/src/test/java/net/squanchy/schedule/FirestoreScheduleServiceTest.kt]: `when onlyFavorites is true` or `when filtering by only favourites`

**Review by rock3r — CHANGES_REQUESTED** [review]
A couple of things to sort out then it's ok for me!

File: app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt (hunk 4 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/service/firestore/FirestoreDbService.kt]: got merged after I did it, but yeah we can do it

**Review by rock3r — APPROVED** [review]

**Pull request merged**